### PR TITLE
Add macOS Support & restructuring for future os support

### PR DESCRIPTION
**Pull request by 123marvin123**
This PR is a big one, since it basically changes the whole structure of the project to allow easier integration with future OS support (e. g. windows). The readouts have been extracted into their own corresponding traits, which allow a cross-platform usage of the api. Each trait can then be implemented on each os.

This PR also adds support for macOS (battery readout not yet finished)

image

**Please don't merge yet, this is just a draft for further discussion.**

### CODE REVIEW

**Comment by 123marvin123** [comment]
image

**Docker Linux** (no Machine name due to Docker?)

image

**macOS Big Sur**

image

**NetBSD**

**Comment by grtcdr** [comment]
Great work @123marvin123 

If Docker Linux doesn't report the product name, we can just fail the `host` component and it'll be automatically hidden, but we'll have to inform the user in the `host` FailedComponent `extraction_method` stating that product name can't be reported on Docker Linux.

Macchina indeed needs some restructuring and I'm glad you did so, I'll review the new changes.

**Comment by grtcdr** [comment]
macOS:
Instead of grouping the machine and OS, you should split them up, like with NetBSD.
It should be:
Machine - iMacPro1,1
OS - macOS 11.2.0 Big Sur

File: src/display.rs [hunk 1]
```diff
@@ -96,6 +97,12 @@ impl Fail {
                 true,
                 String::from("(OK:DISABLED) Distribution -> NetBSD system detected, so the distribution is automatically hidden."),
             ),
+            #[cfg(target_os = "macos")]
+            distro: FailedComponent::new(
+                true,
+                String::from("(OK:DISABLED) Distribution -> macOS system detected, so the \
```
Comment (123marvin123) [inline src/display.rs:103]: @grtcdr Distribution messages for macOS is already here.

**Comment by 123marvin123** [comment]
> ```
> distro: FailedComponent::new(
>                 true,
>                 String::from("oh no :(")
>  ),
> ```
> 
> 
> Do the structural modifications affect `FailedComponent`?
> 
> If they don't, I think this cute "oh no :(" message should be a more comprehensive explanation as to why the `distribution` element failed, again, sort of like with NetBSD, it should say:
> 
> "(OK:DISABLED) Distribution -> macOS system detected, so the distribution is automatically hidden."

No, the failed components will still be hidden. And the "oh no :(" is already replaced in one of the newer commits 😄

File: src/display.rs (hunk 1 above)
Comment (grtcdr) [reply src/display.rs:103]: So when does "oh no" appear?

**Comment by grtcdr** [comment]
Oh, alright, got it!

**Comment by grtcdr** [comment]
This new approach is definitely a lot better than the current one, I also absolutely love the new `bars::memory` refactor you did!

**Comment by 123marvin123** [comment]
> macOS:
> Instead of grouping the machine and OS, you should split them up, like with NetBSD.
> It should be:
> Machine - iMacPro1,1
> OS - macOS 11.2.0 Big Sur

image

How about that? Displays macOS version and kernel version.

**Comment by 123marvin123** [comment]
> Great work @123marvin123
> 
> If Docker Linux doesn't report the product name, we can just fail the `host` component and it'll be automatically hidden, but we'll have to inform the user in the `host` FailedComponent `extraction_method` stating that product name can't be reported on Docker Linux.
> 
> Macchina indeed needs some restructuring and I'm glad you did so, I'll review the new changes.

Docker issue has been resolved and field is hidden when no machine name is available.

**Comment by grtcdr** [comment]
> > macOS:
> > Instead of grouping the machine and OS, you should split them up, like with NetBSD.
> > It should be:
> > Machine - iMacPro1,1
> > OS - macOS 11.2.0 Big Sur
> 
> image
> 
> How about that? Displays macOS version and kernel version.

To have a more consistent feel across platforms, I think the best approach is to also split up the OS and the Kernel.

**Comment by grtcdr** [comment]
I tested your changes and I can say that the program performs noticably better, execution time went from ~42ms down to ~34ms, but there's one issue (on Linux)

Machine name (Product) is empty, and `--debug` states that there are no issues.

**Comment by 123marvin123** [comment]
> > > macOS:
> > > Instead of grouping the machine and OS, you should split them up, like with NetBSD.
> > > It should be:
> > > Machine - iMacPro1,1
> > > OS - macOS 11.2.0 Big Sur
> > 
> > 
> > image
> > How about that? Displays macOS version and kernel version.
> 
> To have a more consistent feel across platforms, I think the best approach is to also split up the OS and the Kernel.

Ok, then I'll have to add another field, because right now the "OS" and "Kernel" field is the same but has a different name depending on the target_os.

**Comment by 123marvin123** [comment]
> I tested your changes and I can say that the program performs noticably better, execution time went from ~42ms down to ~34ms, but there's one issue (on Linux)
> 
> Machine name (Product) is empty, and `--debug` states that there are no issues.

That's really nice, not sure why though, must be general refactoring and removing if-branches? I guess it stacks up.

I'll look at the machine name problem.

**Comment by grtcdr** [comment]
Yeah I guess that could be the cause, I'm reviewing the code as we speak.

**Comment by grtcdr** [comment]
I think I know why this machine name fails to print:
So in the previous implementation, there were four parts that construct the product name:
- `product_name()` -> /sys/class/dmi/id/product_name
- `product_family()` -> /sys/class/dmi/id/product_family
- `product_version()` -> /sys/class/dmi/id/product_version
- `product_vendor()` -> /sys/class/dmi/id/sys_vendor

But your implementation does not take into account the product name:
```
impl ProductReadout for LinuxProductReadout {
    fn new() -> Self {
        LinuxProductReadout
    }

    fn version(&self) -> Result<String, ReadoutError> {
        Ok(extra::pop_newline(fs::read_to_string("/sys/class/dmi/id/product_version")?))
    }

    fn vendor(&self) -> Result<String, ReadoutError> {
        Ok(extra::pop_newline(fs::read_to_string("/sys/class/dmi/id/sys_vendor")?))
    }

    fn family(&self) -> Result<String, ReadoutError> {
        Ok(extra::pop_newline(fs::read_to_string("/sys/class/dmi/id/product_family")?))
    }
}
```

**Comment by 123marvin123** [comment]
> I think I know why the machine name fails to print:
> So in the previous implementation, there were four parts that construct the product name:
> 
> * `product_name()` -> /sys/class/dmi/id/product_name
> * `product_family()` -> /sys/class/dmi/id/product_family
> * `product_version()` -> /sys/class/dmi/id/product_version
> * `product_vendor()` -> /sys/class/dmi/id/sys_vendor
> 
> But your implementation does not take into account the product name:
> 
> ```
> impl ProductReadout for LinuxProductReadout {
>     fn new() -> Self {
>         LinuxProductReadout
>     }
> 
>     fn version(&self) -> Result<String, ReadoutError> {
>         Ok(extra::pop_newline(fs::read_to_string("/sys/class/dmi/id/product_version")?))
>     }
> 
>     fn vendor(&self) -> Result<String, ReadoutError> {
>         Ok(extra::pop_newline(fs::read_to_string("/sys/class/dmi/id/sys_vendor")?))
>     }
> 
>     fn family(&self) -> Result<String, ReadoutError> {
>         Ok(extra::pop_newline(fs::read_to_string("/sys/class/dmi/id/product_family")?))
>     }
> }
> ```

Yes, you are right :) Found this exact problem like 2 minutes ago while debugging. But I already pushed the fix, can you confirm that it works (https://github.com/grtcdr/macchina/pull/23/commits/f524c95ea98f9308caca767b1f40f584aa099586)?

**Comment by grtcdr** [comment]
Yeah, it works now, I'm just going to change `Ok(name)` to `Ok(version)` so it matches the previous implementation and doesn't just print this small string:

Screenshot_20210311_192235

**Comment by grtcdr** [comment]
I'll test Macchina on my NetBSD VM and tell you if there are any issues.

**Comment by 123marvin123** [comment]
> Great, works like a charm now!
> I'll test Macchina on my NetBSD VM and tell you if there are any issues.

Great, I just pushed a fix for a compilation error on bsd. Work fine on my NetBSD VM, please try it out.

**Comment by grtcdr** [comment]
I want to point out that to calculate memory usage, we only need free() and total() in NetBSD's MemoryReadout implementation:
```
impl MemoryReadout for NetBSDMemoryReadout {

    fn new() -> Self {
        NetBSDMemoryReadout
    }

    fn total(&self) -> Result<u64, ReadoutError> { 
        Ok(crate::shared::get_meminfo_value("MemTotal"))
    }

    fn free(&self) -> Result<u64, ReadoutError> {
        Ok(crate::shared::get_meminfo_value("MemFree"))
    }

    fn buffers(&self) -> Result<u64, ReadoutError> { 
        Ok(crate::shared::get_meminfo_value("Buffers"))
    }

    fn cached(&self) -> Result<u64, ReadoutError> { 
        Ok(crate::shared::get_meminfo_value("^Cached"))
    }

    fn reclaimable(&self) -> Result<u64, ReadoutError> { 
        Ok(crate::shared::get_meminfo_value("SReclaimable"))
    }

    fn used(&self) -> Result<u64, ReadoutError> { 
        let total = self.total().unwrap();
        let free = self.free().unwrap();

        Ok(total - free)
    }

}```

Since we don't need `buffers()`, `cached()` or `sreclaimable()`, we can get rid of them.

**Comment by grtcdr** [comment]
I've got some key software missing in my NetBSD VM, so I'm installing those.
But in the meantime, we should work on the documentation.

**Comment by grtcdr** [comment]
Battery information no longer shows up on NetBSD:
Here's envstat -d acpibat0 output:

Screenshot_20210311_214621

Here's `--debug` output:
Screenshot_20210311_214650

**Comment by grtcdr** [comment]
So apparently, the version of ripgrep in NetBSD's repository isn't up to date, and doesn't include PCRE2 engine, something that is required to obtain battery information, I guess we'll leave that to the package maintainers to take care of.

image

**Comment by 123marvin123** [comment]
> So apparently, the version of ripgrep in pkgsrc's repository isn't up to date, and doesn't include PCRE2 engine, something that is required to obtain battery information, I guess we'll leave that to the package maintainers to take care of.
> We should however hide this error message.
> 
> image

I don't know much about NetBSD to be honest, so I guess there's not much we can do right now about that.

**Comment by 123marvin123** [comment]
macOS Battery is now fully working.

WhatsApp Image 2021-03-11 at 22 00 33

**Comment by grtcdr** [comment]
I'm going to test the `--no-messages` that should hide any error messages concerning ripgrep for NetBSD, I'll tell you if hides that error message!

**Comment by grtcdr** [comment]
> > So apparently, the version of ripgrep in pkgsrc's repository isn't up to date, and doesn't include PCRE2 engine, something that is required to obtain battery information, I guess we'll leave that to the package maintainers to take care of.
> > We should however hide this error message.
> > image
> 
> I don't know much about NetBSD to be honest, so I guess there's not much we can do right now about that.

Yeah, I'll contact pin, who is maintaining ripgrep and macchina on pkgsrc and ask if we can solve this ASAP so it doesn't show up on NetBSD users' machines.

**Comment by 123marvin123** [comment]
> I'm going to test the `--quiet` that should hide any error messages concerning ripgrep for NetBSD, I'll tell you if hides that error message!
> 
> Unfortunately, it doesn't :(

What exactly are you executing with the --quiet switch? Try using the "--hide bat" option, this should prevent any call to the ripgrep tool.

**Comment by grtcdr** [comment]
> > I'm going to test the `--quiet` that should hide any error messages concerning ripgrep for NetBSD, I'll tell you if hides that error message!
> > Unfortunately, it doesn't :(
> 
> What exactly are you executing with the --quiet switch? Try using the "--hide bat" option, this should prevent any call to the ripgrep tool.

I was trying to disable the error message from showing up until ripgrep gets updated on pkgsrc

**Comment by grtcdr** [comment]
I removed KernelReadout from format.rs as it was causing a warning.

**Comment by 123marvin123** [comment]
I'll document the traits.

**Comment by 123marvin123** [comment]
image

I added example implementations in the documentation for future implementors to get started.

**Comment by grtcdr** [comment]
> image
> 
> I added example implementations in the documentation for future implementors to get started.

Molto bello!

**Comment by 123marvin123** [comment]
> I'm going to test the `--quiet` that should hide any error messages concerning ripgrep for NetBSD, I'll tell you if hides that error message!
> 
> Unfortunately, it doesn't :(

Have you tried to set the "stderr" property on the command in the rust source code to Stdio::pipe() and then after the executing finished, check if the output in the stderr is not empty. If it is not empty, then the call failed because e.g. there is no battery. If the output is empty, the we can proceed as usual.

https://doc.rust-lang.org/std/process/struct.Command.html#method.stderr

**Comment by grtcdr** [comment]
That's actually a really good idea, I'll update when I get home.

And even if ripgrep gets updated and this error message no longer shows up, it's a good thing to implement so that any users who are rocking an old version of ripgrep don't have to experience this issue.

**Comment by grtcdr** [comment]
I'm going to quickly test Macchina on NetBSD, hopefully nothing goes wrong so I can merge this PR!

**Pull request merged**